File: Terrazas-Chavez-Jose-Alexis/Practicas/C++/practica_4.cpp
/*
Programa que crea un proceso y seguidamente crealiza una sincronización de PID con el proceso
subsecuente, gracias al comando "execl" a través de la libreria "os" , esto quiere decir que el
proceso del padre lo obtendra la efecución del comando ps -f (se coloa f y no fea para visualizar
rapidamente el PID) por lo que en ese momento se detendrá el código de ejecución y no imprime el 
print despues de la sincronización.
*/
#include <iostream>
#include <unistd.h>  // Para execl() y getpid()

using namespace std;

int main() {
    cout << "Proceso padre con PID " << getpid() << endl;

    // Reemplaza el proceso actual con el comando 'ps -f'
    execl("/bin/ps", "ps", "-f", (char*)NULL);

    // Este mensaje nunca se mostrará
    cout << "Nunca me verán" << endl;

    return 0;
}

File: Terrazas-Chavez-Jose-Alexis/Practicas/Python/practica_4.py
'''
Programa que crea tres porcesos y realiza un cambio de contexto cada que uno termina,
el cambio de contexto lo determina el quantum.
Si se quiere detener proceso se tiene que colocar el comando ps -fea en otra terminar 
y una vez ubicado el PID matarlo con el el comando $kill <PID>
'''
import os
import time

def proceso(pid):
    i = 1
    while True:
        for t in range(pid + 1):
            print("\t", end="")  # Imprime una tabulación
        print(f"<{os.getpid()},{i}>")  # Imprime el PID actual y el valor de i
        i += 1
        #time.sleep(1)  # Espera un segundo antes de la siguiente iteración para simular el tiempo de ejecución y se muestre despacio

def main():
    N = 3  # N para crear más o menos procesos
    for i in range(1, N + 1):  # Bucle para crear N procesos
        pid = os.fork()  # Crea un nuevo proceso

        if pid == 0:  # Si es el proceso hijo
            proceso(i)  # Llama a la función "proceso" con el número del proceso
            os._exit(0)  # Sale del proceso hijo después de la ejecución
        else:
            continue  # El proceso padre continúa creando otros hijos

if __name__ == "__main__":
    main()
